### ISSUE TITLE
[leduc] Revisiting Visitors for Modular Extension of DSMLs

### ISSUE THREAD

**Issue by grammarware** [grammarware/models17#2]
Submitted by @tdegueul to https://github.com/grammarware/models17/tree/master/leduc

**Comment by artifacteryx** [comment]
This artifact accompanies a paper on the revisitor pattern, a novel approach to support the non-intrusive extensibility of syntax and semantics definitions of DSLs using object algebras. The artifact includes the implementation of the approach, the evaluation data from the paper, as well as code and instructions for running the benchmark.

(1) Is the artifact consistent with the paper?
 
Yes. For reference, I used Fig. 2 from the paper, which provides a overview of the involved parts. The instructions clearly and reproducibly explain how the two involved code generation processes are triggered. All evaluation artifacts mentioned in the paper are provided within the submission: an activity diagram language meta-model, five solutions for extending the activity diagram with executability, three input models, and the benchmark code for measuring performance.

(2) Is the artifact as complete as possible?

Concerning the formalities of the artifact evaluation process (see https://itu.dk/research/codegrp/), the submission is not yet complete:

- A two-page description document has not been submitted.
- A ReMoDD or FigShare submission is not mentioned.

From the perspective of reproducibility and reuse, the submission seems complete, since all artifacts to reproduce the experiments are contained in the (Github-based) submission. 
 
(3) Is the artifact well-documented?
 
Yes, I was able to reproduce all described steps based on the detailed Readme files.
 
(4) Is the artifact easy to (re)use?
 
- Short-time use and reuse is certainly enabled by the good documentation.
- Long-term use of the artifact is threatened by the use of an update site from an HTTP host (which may be subject to modifications/vanishing). The authors may either make the update site available as part of the artifact, or provide a virtual machine or docker image with a fixed state of the system.

**Comment by jesusc** [comment]
The artifact consists of the code to implement the semantics of fUML
using different approaches to compare their performance against the
revisitor pattern.

(1) Consistent with the paper

The artifact is consistent with the paper. It provides the performance
benchmarks used in the paper as evaluation. I have run the benchmarks
and they are consistent with the results in the paper.

It also provides additional examples.
 
(2) As complete as possible

The artifact is in general complete. However, it could be even more
complete if you provide the raw data that you gathered from the
benchmarks originally (i.e., the csv files).

(3) Well-documented

The documentation could be improved a bit. You should summarize the
artifacts in the README.md and give some details about what is the purpose
of the artifact (to make it more self-contained).

Also, the documentation is scattered through a couple of github
repositories. I suggest that you summarize the instructions in the
artifact evaluation repository to make sure that it is preserved.

(4) Easy to (re)use

The artifact is easy to use. It provides a nice script to launch
the evaluation. The update site and the example code can also be
installed and compiled easily.

The only recommendation is that you could provide a packaged Eclipse
version or a VirtualBox image to make sure that it can always be
reused.

**Comment by patrimrodilla** [comment]
The artefact consists on a new language implementation pattern (REVISITOR) and their evaluation with existing implementation and variants.


(1) Consistent with the paper

Yes. The artifact is well explained and consistent with the paper, including the necesary sources to reproduce the performance evaluation detailed in the paper. I run the benchmarks 



(2) As complete as possible

In general, I think the artefact meets the objectives of demonstration of the language implementation and it advantages. Yo also can run by your own (outside the scripts)  the benchmarks. However, it could be desirable to be able to design your own littel scenario (apart from the TTC'15 etc.) in order to check how well the pattern implementation works with possible own implementation. (I understand that this could be future work...)


(3) Well-documented

I particularly enjoy this aspect in the artefact. It is common that Eclipse-based artefacts are very difficult to reproduce due to the wide broad of versions, plug-ins and dependencies. However, I did not have any problem in reproduce the entire specification and run by my own the benchmarks and also the scripts. Attending to more general documentation (not only for paper evaluation reproduction), maybe could be interesting a more well explained documentation in order to reuse the pattern inside existing projects or existing environments in Eclipse...

(4) Easy to (re)use

I think it is easy to use the Revisitor pattern implementation maintaining the entire environment of the authors. However, I have some concerns about the re(use) aspects out of there (only the language implementation pattern...). I recommend the authors to work in future in modularization and simplification of the solution, in order to transform it in a product-oriented software solution taht could be easily include it in other Java projects.

**Comment by iovinoludovico** [comment]
These artifacts are part of the paper proposing a language implementation pattern that enables independent extensibility of the syntax and semantics of metamodel-based xDSMLs with incremental compilation and without anticipation. 

(1) Is the artifact consistent with the paper?
Yes, the artefacts are consistent with the paper. The authors provided test projects that can be used to demonstrate the applicability of the ALE compiler. The models provided in the examples folder are cited in the paper as running example.

(2) Is the artifact as complete as possible?
Yes the artefacts are complete enough to be executed and tested. I run the ALE generation on the projects verifying the generated code.

(3) Is the artifact well-documented?
Yes the artefact is well documented, since it took me a short time to re-execute the tutorial and understand the places for the generated artefacts. I only have a suggestion: running the generation does not notify when the artefacts are generated, so the authors should indicate where the ale files are placed (folder semantics) or print in the console that the generation is over.

(4) Is the artifact easy to (re)use?
Yes, the artefact it’s easy to reuse and also to extend for the generality of the approach.

**Comment by tdegueul** [comment]
Thank you for your kind and constructive feedback @artifacteryx, @jesusc, @patrimrodilla, and @iovinoludovico. We are pleased to see that you have not encountered major problems in the use of our artifacts.

I will update the artifacts over the weekend to take your remarks into account, and I will keep you updated here.

**Comment by wasowski** [comment]
Dear @tdegueul, 

Thank you for updating many things already.

I do have one technical request: the entire artifact needs to be archived in a single figshare or remodd repository that does not depend on external update  sites, github repositories, etc (including the pdf description file).  

Your current remodd entry seems to link to github repos, etc.  This dependency we should eliminate. It is fine to point to newest version of the tools, but the main purpose of the AE process is to archive exactly the artifact that has accompanied the accepted version of the paper, in a way which is possibly robust to reproduction years later.  So please mirror update sites, zip git sources, and/or pack things into virtual machine images, or do any other similar tricks to make it self contained in robust.  I hope you can get there soon.

**Comment by tdegueul** [comment]
Dear @wasowski,

Our ReMoDD entry (http://www.remodd.org/v1/content/ale-compiler-and-benchmarks) indeed links to the Github repos, but also contains two ZIP files containing everything needed to re-build the artifacts and run the experiments (including READMEs), without any external dependency (no update site, etc.).

I just re-uploaded a new version of the artifact on ReMoDD that includes the PDF description file we created in the meantime. I think the ReMoDD artifact is really self-contained now: source code, build instructions, documentation, evaluation data, results.

Please come back to us if there is any further improvement we should do.

**Comment by wasowski** [comment]
thanks!

**Comment by wasowski** [comment]
Dear Authors

The Artifact Evaluation Committee of MoDELS 2017 has reached the conclusion during the closed committee meeting that this artifact conforms to the expectations and is hereby **approved**. Please use the [badge instructions](https://github.com/grammarware/models17/blob/master/Badge.md) page to add the badge of approval to your article, and add the link to the FigShare entry with a DOI to the camera ready version of the paper.

Thank you very much for putting extra effort into the preparation and finalizing of the artifact. If any of the comments above are still not addressed, please try to accommodate them before the conference.

**Comment by wasowski** [comment]
Sorry for the fishare mention in the repository.  As far as I can see, your artifact is in ReMoDD so you should definitely link to ReMoDD.  Apologies again.

**Comment by tdegueul** [comment]
Thank you for the good news! We've updated the paper with a link to the ReMoDD repository.

As a side note, following the "badge instructions", the badge ends up in an odd area. Is this fine to customize it a bit to fit in a better place?

Original:
![bad-badge](https://user-images.githubusercontent.com/5604095/28573112-cfc62e88-7149-11e7-96db-d3a62bec058b.png)

Customized:
![good-badge](https://user-images.githubusercontent.com/5604095/28573113-cfc69882-7149-11e7-904b-b015d0ab1a57.png)

**Comment by grammarware** [comment]
Absolutely! The badge is tweakable for this precise reason: depending on the title, the number of authors and the length of their affiliations, you might need to move it around.
